Services/Database/classes/Atom/class.ilAtomQuery.php

```diff
@@ -0,0 +1,320 @@
+		self::ISOLATION_READ_UNCOMMITED,
+		self::ISOLATION_READ_COMMITED,
+		self::ISOLATION_REPEATED_READ,
+		self::ISOLATION_SERIALIZABLE,
+	);
+	// Anomalies
+	const ANO_LOST_UPDATES = 1;
+	const ANO_DIRTY_READ = 2;
+	const ANO_NON_REPEATED_READ = 3;
+	const ANO_PHANTOM = 4;
+	/**
+	 * @var array
+	 */
+	protected static $possible_anomalies = array(
+		self::ANO_LOST_UPDATES,
+		self::ANO_DIRTY_READ,
+		self::ANO_NON_REPEATED_READ,
+		self::ANO_PHANTOM,
+	);
+	/**
+	 * @var array
+	 */
+	protected static $anomalies_map = array(
+		self::ISOLATION_READ_UNCOMMITED => array(
+			self::ANO_LOST_UPDATES,
+			self::ANO_DIRTY_READ,
+			self::ANO_NON_REPEATED_READ,
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_READ_COMMITED   => array(
+			self::ANO_NON_REPEATED_READ,
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_REPEATED_READ   => array(
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_SERIALIZABLE    => array(),
+	);
+	/**
+	 * @var int
+	 */
+	protected $isolation_level = self::ISOLATION_SERIALIZABLE;
+	/**
+	 * @var array
+	 */
+	protected $tables = array();
+	/**
+	 * @var callable
+	 */
+	protected $query;
+	/**
+	 * @var Closure[]
+	 */
+	protected $queries = array();
+	/**
+	 * @var
+	 */
+	protected $ilDBInstance;
+	/**
+	 * @var ilAtomQuery
+	 */
+	protected static $instance;
+
+
+	/**
+	 * ilAtomQuery constructor.
+	 *
+	 * @param \ilDBInterface $ilDBInstance
+	 * @param int $isolation_level currently only ISOLATION_SERIALIZABLE is available
+	 */
+	public function __construct(ilDBInterface $ilDBInstance, $isolation_level = self::ISOLATION_SERIALIZABLE) {
+		$this->ilDBInstance = $ilDBInstance;
+		$this->isolation_level = $isolation_level;
+	}
+
+	//
+	//
+	//
+	/**
+	 * @return array
+	 */
+	public function getRisks() {
+		return static::getPossibleAnomalies($this->getIsolationLevel());
+	}
+
+
+	/**
+	 * Add table-names which are influenced by your queries, MyISAm has to lock those tables. Lock
+	 *
+	 * @param string $table_name
+	 * @param int $lock_level use ilAtomQuery::LOCK_READ or ilAtomQuery::LOCK_WRITE
+	 * @throws \ilDatabaseException
+	 */
+	public function addTable($table_name, $lock_level) {
+		if (!in_array($lock_level, array( static::LOCK_READ, static::LOCK_WRITE ))) {
+			throw new ilDatabaseException('The current Isolation-level does not support the desired lock-level. use ilAtomQuery::LOCK_READ or ilAtomQuery::LOCK_WRITE');
+		}
+		$this->tables[] = array( $table_name, $lock_level );
+	}
+
+
+	/**
+	 * Every action on the database during this isolation has to be passed as Closure to ilAtomQuery.
+	 * An example:
+	 * $ilAtomQuery->addQueryClosure( function ($ilDB) use ($new_obj_id, $current_id) {
+	 *        $ilDB->manipulateF('
+	 *            DELETE FROM frm_user_read
+	 *            WHERE obj_id = %s AND thread_id =%s',
+	 *        array('integer', 'integer'),
+	 *        array($new_obj_id, $current_id));
+	 *    });
+	 *
+	 * @param \Closure $query
+	 */
+	public function addQueryClosure(Closure $query) {
```

Comment:
I would prefer if we would stick to the `Closure`. `Callable` introduces all kinds of weirdness due to strings or arrays interpreted as names of some function or method. IMO that's legacy from PHPs wild teenage years and will be removed some times soon, as PHP is becoming a grown up now. If anyone really needs it, one could wrap the callable in a closure and use the API anyway. The weirdness is then at least contained at the side of the caller.